Master-Slave Aggregating Relay (multiple master setup)

greenbender opened the pull request:
This pull request adds support for multiple Master-Slave relays. Something I am calling a Master-Slave Aggregating Relay.

A couple of people have asked for this feature in the past and were referred to the Single-Broadcast Relay, which is OK provided you know the specific mountpoints that you want to relay ahead of time.

If you are in a position where you want to relay mountpoints for more than one the master server, and one or more of the master servers adds and removes mountpoints dynamically or if you are unable to know what mountpoints are available ahead of time then this gives you the ability to aggregate all mountpoints for any numbers of master servers.

Namespacing is also supported to prevent name clashes for mountpoints being relayed from different servers.

ePirat commented:
Please do not include the `Revert "Changed URLs to be relative in .gitmodules"` commit in your pull request.

ePirat commented:
So, sorry for getting right to the point on my first comment, it was just one thing that I noticed immediately. (If it's too much hassle to remove it, don't worry, I can do that manually later)

First of all, thanks a lot for the pull request! It needs some review and discussion and, of course, some testing before we can decide if we can merge it.

ePirat commented on src/slave.c:
```diff
@@ -767,10 +789,18 @@ static void *_slave_thread(void *arg)
+             * the master linked list and releasing the lock */
+            master = config->master;
+            while (master) {
+                update_from_master(master);
```
This seems indeed problematic to do, as it can hold the config lock for quite some time, if fetching the streamlist takes a while or you have a lot of master servers.

greenbender replied:
OK ... Any recommended approach?

I'll start by copying the master linked list out of the config and releasing the lock, then iterate the copied linked list.

Approaching it in a similar manner to the relay linked list might be preferable though, in that the copied master linked list (eg. global.master) would only have to be updated when the config changes.

greenbender commented:
Removed that `Revert "Changed URLs to be relative in .gitmodules"` commit.

I have also implemented the copy master servers list from config change to minimize config locking.

ePirat commented:
> I have also implemented the copy master servers list from config change to minimize config locking.

Thanks, I think this is better now. Will still need to discuss this with the other project members which can take a while, just wanted to inform you, so that you know that I haven't forgotten your PR. :smile:

LogicalNetworkingSolutions commented:
@greenbender - with this patch would master and slave servers have different named mountpoints (such as with a configured namespace of "upstream1-",  /mount.opus on master would be "/upstream1-mount.opus" on slave server(s) ? This is an incredibly useful feature addition IMHO and look forward to using it =} Thank you.

greenbender commented:
At the moment the namespace is simply pre-pended to the mountpoint so in your example your might make the namespace "/upstream1" (yes a leading forward slash - thinking of making that optional?) and so the mountpoint "/mount.opus" on the master would become "/upstream1/mount.opus" on the slave.

The "mesh" sounds more like a use case where all streams would have to been uniquely named and you wouldn't use a namespace (If I understand correctly) then all servers would have a complete set of mountpoints for the rest of the "mesh" - Its been a while since I wrote this (I'm not sure what happens when a mountpoint path is already in use, does it get ignored/error/override). Perhaps I'm misunderstanding your requirements?

LogicalNetworkingSolutions commented:
Oh, I see how it works ("/upstream1/mount.opus", not with a hyphen). That makes a lot of sense. I do know that, without master-slave aggregation when you try to make a master and slave relay each other with duplicate mountpoint names, weird things happen(tm). I think you answered my question with your first comment regarding the pathname of the aggregated mountpoint. Seems like listener clients wouldn't have a problem digesting a nested mountpoint like this? Thanks for your response =}

ePirat commented:
> Also, seems like listener clients wouldn't have a problem digesting a nested mountpoint like this? 

That shouldn't be a problem for clients, as it's just a HTTP resource to them.

ePirat commented on src/slave.c:
```diff
@@ -63,6 +63,103 @@ static volatile int update_all_mounts = 0;
+
+
+/* free master list and return NULL */
+static master_server *master_list_free(master_server *list)
```
I would not return NULL here but make it void. It is a smart idea but personally I think it makes code that comes later where you do:
```
foo = master_list_free(foo);
```
a bit confusing if one is not familiar with this behavior of `master_list_free`.

greenbender replied:
I have implemented this change

greenbender commented on src/cfgfile.c:
```diff
@@ -1579,6 +1588,76 @@ static void _parse_http_headers(xmlDocPtr                   doc,
+
+    master->next         = NULL;
+    master->on_demand    = configuration->on_demand;
+    master->server       = (char *) xmlCharStrdup("127.0.0.1");
```
It seemed like a reasonable default at the time ```master->server = NULL``` would also be fine though it should perhaps be logged as a bad config if ```master->server``` is not set.

sanduluca commented:
Why there is not activity on this enhancement ? Would be great to have this feature.

luzpaz commented:
Does this need to be transferred to gitlab ?

shaksp commented:
@greenbender thanks for this - I am looking for this feature as well but it's not yet been merged. I note your source is built on 2.4.1. Do you happen to have a version of this based on 2.4.4/2.4.5 that I could use and build from source? Many Thanks!
